lib/service/chat_service.dart

```diff
@@ -0,0 +1,213 @@
+
+    // listen events
+
+    _websocketChannel?.stream.listen(
+      (event) {
+        log.info("[CHAT] event: $event");
+        final response = app.WebsocketMessage.fromJson(json.decode(event));
+        switch (response.command) {
+          case 'NEW_MESSAGE':
+            try {
+              final newMessages = (response.payload["messages"]
+                      as List<dynamic>)
+                  .map((e) => app.Message.fromJson(e as Map<String, dynamic>))
+                  .toList();
+              for (var element in _listeners) {
+                element.onNewMessages(newMessages);
+              }
+            } catch (e) {
+              log.info("[CHAT] NEW_MESSAGE error: $e");
+            }
+            break;
+          case 'RESP':
+            if (response.payload["ok"] != null &&
+                response.payload["ok"].toString() == "1") {
+              for (var element in _listeners) {
+                element.onResponseMessage(response.id, ChatService.SENT);
+              }
+            } else if (response.payload["error"] != null) {
+              for (var element in _listeners) {
+                element.onResponseMessage(response.id, ChatService.ERROR);
+              }
+            } else {
+              try {
+                final newMessages =
+                    (response.payload["messages"] as List<dynamic>)
+                        .map((e) => app.Message.fromJson(e))
+                        .toList();
+                for (var element in _listeners) {
+                  element.onResponseMessageReturnPayload(
+                      newMessages, response.id);
+                }
+              } catch (e) {
+                log.info("[CHAT page] RESP error: $e");
+              }
+            }
+            break;
+          default:
+            break;
+        }
+      },
+      onDone: () async {
+        log.info("[CHAT] onDone");
+        if (_listeners.isEmpty) return;
+        Future.delayed(const Duration(seconds: 5), () async {
+          if (_address != null && _id != null && _wallet != null) {
+            _address = null;
+            _id = null;
+            _wallet = null;
+            log.info("[CHAT] _websocketChannel reconnecting");
+            await connect(
+              address: _address!,
+              id: _id!,
+              wallet: _wallet!,
+            );
+            for (var element in _listeners) {
+              element.onDoneCalled.call();
+            }
+          }
+        });
+      },
+    );
+  }
+
+  @override
+  void addListener(ChatListener listener) {
+    _listeners.add(listener);
+  }
+
+  @override
+  Future<void> dispose() async {
+    log.info("[CHAT] disconnect");
+    _address = null;
+    _id = null;
+    _wallet = null;
+    await _websocketChannel?.sink.close();
+    _websocketChannel = null;
+  }
+
+  @override
+  void sendMessage(dynamic message) {
+    _websocketChannel?.sink.add(message);
+  }
+
+  Future<Map<String, dynamic>> _getHeader(
```

Comment:
This is better to be in `HttpHelper`

Reply:
message formats of each api are too different to be reuse, so it's okay to keep it private

Reply:
they should be the same for all hmac.

Reply:
that depends on api design, for instance, some place use timestamp is in second, some is in millisecond

Reply:
No, as confirmed from Jim before, the Hmac must be the same for all Autonomy Apis.

Reply:
so it has to be fix from server side first. we must have atleast the document first, then implement later.

Reply:
No i mean server is now using the same hmac logic for all api, and client should do too.
See https://hackmd.io/@bitmarkinc/SkxGvZ3W3?type=view